### PR TITLE
[WIP] Port to Fabric

### PR DESCRIPTION
**Pull request by hanbings**
https://github.com/WiIIiam278/HuskSync/issues/117

**There is still a lot of work to be done. This PR is to keep in sync with the upstream code. Review work needs to be postponed a little longer. Sorry for taking so much time.**

## Basic Information:

Target game version: 1.20.1
Fabric Loader version: 0.14.21
Fabric API version: 0.83.0+1.20.1

## Progress:

**Basic**

- [x] build system
- [x] configuration file.

**Database**

- [x] Mysql
- [x] Redis

**Events**

- [x] player joins the game
- [x] player quit the game
- [x] player death
- [x] server saved world
- [ ] projectile
- [ ] player drop
- [ ] player pick
- [ ] player interact
- [ ] player place block
- [ ] player break block
- [ ] open the Inventory
- [ ] hurt
- [ ] command

**Data**

- [ ] Serialization and Deserialization
- [ ] Code for Platform Setup and Get Data Objects (FabricPlayer.class) *partially done*

**API** *No plans yet*

**Migration tool** *No plans yet*

**Analytics**

- [x] Plan Analytics

## Currently encountered problems:

1. The current Bukkit platform code uses Bukkit's Serializer tool, which has some special behaviors for Bukkit serializable classes, which are difficult to implement on the Fabric platform.
2. There is a lot of code for Setter Getter behavior in a partial class file, considering using Lombok to simplify the code? (this is just my personal suggestion)
3. Fabric Loom requires Gradle 8.0 or higher, currently upgraded to 8.1.1.

### CODE REVIEW

File: fabric/src/main/java/net/william278/husksync/FabricHuskSync.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+            // Register permissions
+            log(Level.INFO, "Registering permissions & commands...");
+
+            // Register commands handler
+            log(Level.INFO, "Successfully registered permissions & commands");
+
+            // FIXME: The plan function is not supported.
+            // FIXME: The bStats metrics function is not supported.
+            // FIXME: The updates function is not supported.
```
Comment (WiIIiam278) [inline fabric/src/main/java/net/william278/husksync/FabricHuskSync.java:150]: Totally fine to only support certain features on certain platforms where there's no equivalent.

I think Plan is supported on Fabric? Definitely okay to deprioritize some of this stuff.

**Comment by WiIIiam278** [comment]
Breathtaking work thus far.  :+1:

**Comment by hanbings** [comment]
Thank you so much for your endorsement! Next I'll peruse the source code of Bukkit serialization for the correct way they work. There will be new commits here soon, once again sorry for the speed of work.

File: fabric/src/main/java/net/william278/husksync/FabricHuskSync.java (hunk 1 above)
Comment (hanbings) [reply fabric/src/main/java/net/william278/husksync/FabricHuskSync.java:150]: I overlooked this, Plan is statistically supported by Fabric, I will check their documentation as soon as possible.

**Comment by WiIIiam278** [comment]
All good! Thanks for your work. Just a heads up -- I've changed the license to Apache-2.0 today to be more permissive -- I hope this is a bit better to work with! :)

I hope that's okay with you? You'll need to update it to Apache-2.0 in the fabric mod config, too :)

**Comment by hanbings** [comment]
My pleasure, Thanks for contributing to open source!

**Comment by Hlebuw3k** [comment]
Hey this is awesome I need this

**Comment by Hlebuw3k** [comment]
I was looking at the code, and if the player is in adventure game mode, does it change to survival game mode?
fabric/src/main/java/net/william278/husksync/player/FabricPlayer.java
line 77

**Comment by hanbings** [comment]
> I was looking at the code, and if the player is in adventure game mode, does it change to survival game mode? fabric/src/main/java/net/william278/husksync/player/FabricPlayer.java line 77

It does seem that Adventure mode is not recognized. I’ve been very busy at work recently, so I haven’t updated it for a long time: (. If you are willing, you are welcome to PR!

**Comment by MasterSwish** [comment]
> Hey this is awesome I need this

Me too! Thank you for your work on this!

**Comment by WiIIiam278** [comment]
Probably a good idea to rebase this on the new serialisation-refactor branch when that’s merged

**Comment by xPand4B** [comment]
I'd love to use this mod on our fabric network!

@hanbings @WiIIiam278 So you need some coding support for this one?

**Comment by hanbings** [comment]
> Probably a good idea to rebase this on the new serialisation-refactor branch when that’s merged as that’ll help big time with a few of the todos on this PR. Also, new task system can borrow HuskHomes existing fabric code

Do you mean to merge this PR first and then fork it for the next step? Of course, it's up to you.

**Comment by hanbings** [comment]
> I'd love to use this mod on our fabric network!
> 
> @hanbings @WiIIiam278 So you need some coding support for this one?

Yes. There has been no progress for a long time due to my work. It's currently hard for me to find the time to work on Bukkit's item deserialization, which will result in us not being able to correctly parse the player's item information from the database.

**Comment by WiIIiam278** [comment]
@hanbings No problem at all!
I'm currently refactoring the plugin on the serialization-refactor branch to use NBTAPI instead of Bukkit serialisation, so this should make this PR much easier to finish up when done.

**Comment by hanbings** [comment]
> @hanbings No problem at all! I'm currently refactoring the plugin on the serialization-refactor branch to use NBTAPI instead of Bukkit serialisation, so this should make this PR much easier to finish up when done.

Thank you for your attention! I've synced the upstream code. Now that I'm ready to merge the code I've completed so far (this PR), I'll compile it in my local environment later to make sure the fabric version can pass the build.

**Comment by hanbings** [comment]
Actually I have some doubts why here (https://github.com/WiIIIam278/HuskSync/blob/master/bukkit/src/main/java/net/william278/husksync/BukkitHuskSync.java#L99) uses a boolean as the loading success flag bit. Would it be better here to catch exceptions thrown from the jdbc driver?

**Comment by WiIIiam278** [comment]
Agreed — I suggest merging with the https://github.com/WiIIiam278/HuskSync/tree/v3-work branch which already fixes this

**Comment by WiIIiam278** [comment]
v3.0, just merged, should make implementing this a lot easier. happy to help work on that if you're interested :)

**Comment by WiIIiam278** [comment]
yep :)

**Comment by WiIIiam278** [comment]
@hanbings I'm going to commit to this branch if it's okay to help you out with this -- please let me know if there's anything you'd prefer to work on / feel free to self-review the PR. Thanks very much!

**Comment by hanbings** [comment]
> @hanbings I'm going to commit to this branch if it's okay to help you out with this -- please let me know if there's anything you'd prefer to work on / feel free to self-review the PR. Thanks very much!

It seems that there is a big difference between 3.0 and previous versions. I am not sure what to do next. My current idea is to manually modify the previous code and then determine our new progress. What do you think?

And... thanks for your continued follow-up!

**Comment by WiIIiam278** [comment]
> > @hanbings I'm going to commit to this branch if it's okay to help you out with this -- please let me know if there's anything you'd prefer to work on / feel free to self-review the PR. Thanks very much!
> 
> It seems that there is a big difference between 3.0 and previous versions. I am not sure what to do next. My current idea is to manually modify the previous code and then determine our new progress. What do you think?
> 
> And... thanks for your continued follow-up!

No problem! I have done a fair bit of work today. The next thing to do is implement Data and Serializable for each class type. Minecraft provides utils for getting NBT from Items, so we can do that. Persistent Data Containers isn't a feature in the Vanilla game, so this will utilize the custom data store I wrote for third party plugin/mod support. 

And a bunch of mixins to handle events -- this is something I'm not very good at, though!

**Comment by WiIIiam278** [comment]
@hanbings To keep you in the loop, I've added Data and Serializer classes for Item data types to NBT. Needs testing, but just need to add equivalent ones for the other data types and the base sync system should work :)

The big remaining work is mixin logic for handling the various events that don't have workable API equivalents.

**Comment by WiIIiam278** [comment]
I'm still slowly chipping away at this bit by bit and awhile ago added Fabric implementations of inventory data types.

If anyone more experienced with mixins is able to have a look at some of the needed ones for this PR, get in touch!

**Comment by Stampede2011** [comment]
I have begun implementing more of the events needed for this, [here](https://github.com/Stampede2011/HuskSync). I am still working on ensuring each event is precise, but the basics are there.

- [ ] projectile
- [x] player drop (needs more testing)
- [x] player pickup
- [x] player interact
- [x] player place block
- [x] player break block
- [ ] open the Inventory
- [x] hurt
- [x] command

**Comment by WiIIiam278** [comment]
@Stampede2011 Hey, great work! Feel free to PR into this branch or even into master if you want to supersede this PR :)

**Comment by Stampede2011** [comment]
> @Stampede2011 Hey, great work! Feel free to PR into this branch or even into master if you want to supersede this PR :)

Just created a PR for this if your able to review: https://github.com/hanbings/HuskSync/pull/1

**Comment by WiIIiam278** [comment]
> > @Stampede2011 Hey, great work! Feel free to PR into this branch or even into master if you want to supersede this PR :)
> 
> Just created a PR for this if your able to review: [hanbings#1](https://github.com/hanbings/HuskSync/pull/1)

Had a quick cursory glance just now, and this looks to be excellent work.

What I'm going to do, in the interest of making this a bit easier for me to manage and review, is merge this PR into a `feat/fabric` branch on this repository. I'll open another tracking PR for merging that branch into master, at which point you can target your PR to merge into the `feat/fabric` branch :)

**Comment by WiIIiam278** [comment]
Superseded by #217

**Pull request closed without merging**